src/Onnx/OnnxToFeld.hs

```diff
@@ -236,42 +264,142 @@ mkInitReadFile hf initGroups = L.concat $ start : map initVec initGroups ++ map
 tiToType :: TensorInfo -> Type
-tiToType ti = goDim (D.length $ tiDims ti) (toType $ tiType ti)
+tiToType ti = tensorToType (D.length $ tiDims ti) (tiType ti)
+
+-- | Compute the (Program) type corresponding to a tensor with the given dimensionality and element type
+tensorToType :: Int -> TD.DataType -> Type
+tensorToType n t = goDim n $ toType t
   where goDim 0 _  = error "Zero dimensional tensor not implemented"
         goDim 1 tt = mkAwLType universal tt
-        goDim n tt = mkStructType
+        goDim m tt = mkStructType
                        [("member1", mkAwLType universal tt),
-                        ("member2", mkStructType $ map dimField [1 .. n])
+                        ("member2", mkStructType $ map dimField [1 .. m])
                        ]
-        dimField n = ("member" ++ show n, 1 :# NumType Unsigned S32)
+        dimField i = ("member" ++ show i, 1 :# NumType Unsigned S32)
 
 -- | Compute code for initializing one component of the weight record
 initTensor :: TensorInfo -> L.ByteString
-initTensor ti = (<>) "\n" $ L.unlines $ map ("  " <>)
-              $ [readDims] ++ initSizes dims ++ [alloc awlPrefix] ++ readElems awlPrefix
+initTensor ti = allocReadTensor "wfile" prefix (D.length $ tiDims ti) (tiType ti)
   where prefix = "w->member" <> show (tiField ti + 1) <> ".buffer[" <> show (tiIdx ti) <> "]"
-        awlPrefix = if length dims > 1 then prefix <> ".member1" else prefix
-        elemTH = showElemT elemTT
+
+-- | Allocate memory for the elements of a tensor and read the contents from a file
+allocReadTensor :: L.ByteString -> L.ByteString -> Int -> TD.DataType -> L.ByteString
+allocReadTensor file prefix nDims elemTT = L.unlines code <> "\n"
+  where code = map ("  " <>) $ readDims ++ [alloc] ++ readElems
         elemTC = showCElemT elemTT
-        elemTT = tiType ti
-        dims = D.toList $ tiDims ti
-        initSizes []  = error "Zero dimensional tensor not implemented"
-        initSizes [d] = [awlPrefix <> ".length = " <> show d <> ";"]
-        initSizes ds  = [ prefix <> ".member2.member" <> show j <> " = " <> show n <> ";"
-                        | (n,j) <- zip ds [1 :: Int ..]
-                        ] ++ [awlPrefix <> ".length = " <> show (product dims) <> ";"]
-        readDims = "fscanf(wfile, \" " <> elemTH <> " " <> show (length dims) <> " " <> L.unwords (map show dims) <> "\");"
-        alloc awl = awl <> ".buffer = malloc(" <> awl <> ".length * sizeof(" <> elemTC <> "));"
-        readElems awl = [ "for(int i = 0; i < " <> awl <> ".length; i++) {"
-                        , "  fscanf(wfile, " <> scanFormat elemTT <> ", " <> awl <> ".buffer + i);"
-                        , "}"
-                        ]
+        (awlPrefix, readDims) = initTensorDims file prefix nDims
+        alloc = awlPrefix <> ".buffer = malloc(" <> awlPrefix <> ".length * sizeof(" <> elemTC <> "));"
+        readElems = [ "for(int i = 0; i < " <> awlPrefix <> ".length; i++) {"
```

Comment:
Isn't `++i` easier to generate instructions for if it works?

Space between for and the opening parenthesis.

Reply:
I think there is only a performance difference between ++i and i++ if you use it in an expression; in that case you avoid a copy in case of ++i.